File: 2618-maximize-the-minimum-powered-city/2618-maximize-the-minimum-powered-city.go
type Solver struct {
	n    int
	r    int
	k    int64
	pref []int64
}

func NewSolver(stations []int, r int, k int) *Solver {
	n := len(stations)
	pref := make([]int64, n)
	pref[0] = int64(stations[0])
	for i := 1; i < n; i++ {
		pref[i] = pref[i-1] + int64(stations[i])
	}
	return &Solver{n, r, int64(k), pref}
}

func (s *Solver) check(target int64) bool {
	added := make([]int64, s.n)
	extraSum := int64(0)
	k := s.k

	for i := 0; i < s.n; i++ {
		leftIdx := max(i-s.r-1, 0)
		rightIdx := min(i+s.r, s.n-1)
		var left int64
		if i-s.r-1 >= 0 {
			left = s.pref[leftIdx]
		} else {
			left = 0
		}
		right := s.pref[rightIdx]
		extraSum -= added[leftIdx]
		sum := right - left + extraSum

		diff := target - sum
		if diff > 0 {
			if k-diff < 0 {
				return false
			}
			added[rightIdx] += diff
			extraSum += diff
			k -= diff
		}
	}

	return true
}

func (s *Solver) solve() int64 {
	var res int64
	lo, hi := int64(0), s.pref[s.n-1]+s.k
	for lo <= hi {
		mid := (lo + hi) >> 1
		if s.check(mid) {
			res = mid
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}

	return res
}

func maxPower(stations []int, r int, k int) int64 {
	s := NewSolver(stations, r, k)

	return s.solve()
}

File: 2618-maximize-the-minimum-powered-city/2618-maximize-the-minimum-powered-city.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
class Solution {
   public:
    int n;
    bool solve(ll target, vector<ll>& pref, int r, ll k) {
        vector<ll> added(n, 0);

        ll extrasum = 0;
        for (int i = 0; i < n; i++) {
            int leftIdx = max(i - r - 1, 0);
            int rightIdx = min(n - 1, i + r);
            ll left = i - r - 1 >= 0 ? pref[i - r - 1] : 0;
            ll right = pref[rightIdx];
            extrasum -= added[leftIdx];
            ll sum = right - left + extrasum;

            ll diff = target - sum;
            if (diff > 0) {
                if (k - diff < 0) return false;
                added[rightIdx] += diff;
                extrasum += diff;
                k -= diff;
            }
        }

        return true;
    }
    long long maxPower(vector<int>& stations, int r, int k) {
        this->n = stations.size();
        vector<ll> pref(n, 0);
        pref[0] = stations[0];
        for (int i = 1; i < n; i++) {
            pref[i] = pref[i - 1] + stations[i];
        }
        ll lo = 0, hi = pref.back() + k, ans = 0;

        while (lo <= hi) {
            ll mid = (lo + hi) >> 1;

            if (solve(mid, pref, r, k)) {
                ans = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return ans;
    }
};